packages/ui/src/components/filters/manage-views.tsx

```diff
@@ -0,0 +1,225 @@
+  Button,
+  Icon,
+  Input
+} from '@/components'
+import { DndContext } from '@dnd-kit/core'
+import { SortableContext, useSortable, verticalListSortingStrategy } from '@dnd-kit/sortable'
+import { CSS } from '@dnd-kit/utilities'
+
+import useDragAndDrop from '../../hooks/use-drag-and-drop'
+import { SavedView } from './types'
+
+interface SortableItemProps {
+  view: SavedView
+  id: string
+  onRename: (id: string, newName: string) => void
+  onDelete: (id: string) => void
+  existingNames: Set<string>
+}
+
+const SortableItem = memo<SortableItemProps>(({ view, id, onRename, onDelete, existingNames }) => {
+  const [name, setName] = useState(view.name.trim())
+  const [hasError, setHasError] = useState(false)
+
+  // Hook from dnd-kit for drag and drop functionality
+  const { attributes, listeners, setNodeRef, transform, transition } = useSortable({ id })
+
+  /**
+   * Handles input changes and validates the new name
+   * Checks for duplicates and empty values
+   * Updates parent component state regardless of validation to properly handle Save button state
+   */
+  const handleChange = useCallback(
+    (e: ChangeEvent<HTMLInputElement>) => {
+      const newName = e.target.value
+      setName(newName)
+
+      const trimmedName = newName.trim().toLowerCase()
+      // Check if name exists and is not the current view's name
+      const exists = trimmedName && existingNames.has(trimmedName) && trimmedName !== view.name.trim().toLowerCase()
+      const hasValidationError = exists || !trimmedName
+
+      setHasError(hasValidationError)
+      // Always update parent to reflect current state
+      onRename(view.id, newName)
+    },
+    [view.id, view.name, existingNames, onRename]
+  )
+
+  return (
+    <div
+      ref={setNodeRef}
+      style={{ transform: CSS.Transform.toString(transform), transition }}
+      className="flex items-center gap-x-2 p-2 hover:bg-background-4 rounded"
+    >
+      <div className="cursor-move group" {...attributes} {...listeners}>
+        <Icon
+          className="text-icons-4 group-hover:text-icons-2 transition-colors duration-200"
+          name="grid-dots"
+          size={14}
+        />
+      </div>
+
+      <Input
+        theme={hasError ? 'danger' : 'default'}
+        value={name}
+        size="md"
+        placeholder="Enter a name"
+        onChange={handleChange}
+      />
+
+      <button
+        className="text-icons-4 hover:text-icons-danger"
+        onClick={() => onDelete(view.id)}
+        aria-label="Delete view"
+      >
+        <Icon className="rotate-45" name="plus" size={14} />
+      </button>
+    </div>
+  )
+})
+
+SortableItem.displayName = 'SortableItem'
+
+interface ManageViewsProps {
+  open: boolean
+  onOpenChange: (open: boolean) => void
+  views: SavedView[]
+  onSave: (views: SavedView[]) => void
+  onRename: (id: string, newName: string) => void
+}
+
+const ManageViews: FC<ManageViewsProps> = memo(({ open, onOpenChange, views, onSave }) => {
```

Comment:
this contains a LOT of business logic, which we decided to avoid in `ui` package.

we want to follow the MVVM pattern, where the "view" layer, should only be the presentational layer.

we want to be able to re-use the views across two different backends, which is not possible if we hard-code business logic in the view layer.

Reply:
Here, all the business logic has also been moved to the [use-view-management](https://github.com/harness/canary/pull/513/files#diff-c58401d04ae9a913353d34aeabc42330af54d16582e87e43325c039034e9c35bR43) hook at the page level, which is used for validating and managing all the local states of this component.